pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -659,10 +663,49 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 // GetDeviceRunContainerOptions checks whether we have cached containerDevices
 // for the passed-in <pod, container> and returns its DeviceRunContainerOptions
 // for the found one. An empty struct is returned in case no cached state is found.
-func (m *ManagerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) *DeviceRunContainerOptions {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-	return m.podDevices.deviceRunContainerOptions(string(pod.UID), container.Name)
+func (m *ManagerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) (*DeviceRunContainerOptions, error) {
+	podUID := string(pod.UID)
+	contName := container.Name
+	for k, v := range container.Resources.Limits {
+		resource := string(k)
+		needed := int(v.Value())
+		glog.V(3).Infof("needs %d %s", needed, resource)
+		_, registeredResource := m.healthyDevices[resource]
+		_, allocatedResource := m.allocatedDevices[resource]
+		// Continues if this is neither an active device plugin resource nor
+		// a resource we have previously allocated.
+		if !registeredResource && !allocatedResource {
+			continue
+		}
+
+		m.mutex.Lock()
+		e, ok := m.endpoints[resource]
+		if !ok {
+			m.mutex.Unlock()
+			return nil, fmt.Errorf("endpoint not found in cache for a registered resource: %s", resource)
```

Comment:
This seems a behavior change in that even though the device plugin doesn't require PreStart, we will still fail to restart the container. Instead of making invokePreStartBoolFlag an endpoint field, can we add "pluginOpts map[string]pluginapi.DevicePluginOptions" in ManagerImpl? We can update this cached data when addEndpoint but don't remove the resource when an endpoint stops.